### PR TITLE
Allow depending on a component to mean you depend on it's transitively reachable custom resource children.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
https://github.com/pulumi/pulumi/pull/2494

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@lukehoban @swgillespie @pgavlin PTAL.  This incorporates our new formalization that only 'component resources' are used to aggregate other resources.

File: sdk/nodejs/tests/runtime/langhost/cases/023.parent_child_dependencies_3/index.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// Test the ability to invoke provider functions via RPC.
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply sdk/nodejs/tests/runtime/langhost/cases/023.parent_child_dependencies_3/index.js]: thanks!

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,12 +1,18 @@
 ## 0.16.18 (Unreleased)
 
 ### Improvements
-
 - Attempting to convert an [Output<T>] to a string or to JSON will now result in a warning
   message being printed, as well as information on how to rectify the situation.  This is
   to help with diagnosing cryptic problems that can occur when Outputs are accidentally
   concatenated into a string in some part of the program.
 
+- Depending on a **Component** Resource will now depend on all other Resource's parented by that
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply CHANGELOG.md]: fixed.

File: sdk/nodejs/resource.ts [hunk 3]
```diff
@@ -29,6 +29,52 @@ export abstract class Resource {
...
+     *     var c1 = new CustomResource("c1");
+     *     var c2 = new CustomResource("c2", { parentId: c1.id }, { parent: c1 });
+     *     var c3 = new CustomResource("c3", { parentId: c1.id }, { parent: c1 });
+     * ```
+     *
+     * The problem here is that 'c2' has a data dependency on 'c1'.  If it tries to wait on 'c1' it
+     * will walk to the children and wait on them.  This will mean it will wait on 'c3'.  But 'c3'
+     * will be waiting in the same manner on 'c2', and a cycle forms.
+     *
+     * This cannot happen with ComponentResources as they do not have any data flowing into them.
```
Comment (pgavlin) [inline sdk/nodejs/resource.ts]: This isn't strictly true: components can still participate in inputs/outputs from `registerResource` just like custom resources. However, that scenario is so obscure that I think that it's safe to assume that it will never happen.

File: sdk/nodejs/resource.ts [hunk 4]
```diff
@@ -48,7 +94,7 @@ export abstract class Resource {
     /* @internal */ private readonly __providers: Record<string, ProviderResource>;
 
     public static isInstance(obj: any): obj is Resource {
-        return obj && obj.__pulumiResource;
+        return obj && (<Resource>obj).__pulumiResource === true;
```
Comment (pgavlin) [inline sdk/nodejs/resource.ts]: Why was this change necessary? Just for consistency?

File: sdk/nodejs/resource.ts [hunk 5]
```diff
@@ -211,7 +261,7 @@ export abstract class CustomResource extends Resource {
      * multiple copies of the Pulumi SDK have been loaded into the same process.
      */
     public static isInstance(obj: any): obj is CustomResource {
-        return obj && obj.__pulumiCustomResource;
+        return obj && (<CustomResource>obj).__pulumiCustomResource === true;
```
Comment (pgavlin) [inline sdk/nodejs/resource.ts]: Same here

File: sdk/nodejs/runtime/resource.ts [hunk 6]
```diff
@@ -258,17 +260,35 @@ async function prepareResource(label: string, res: Resource, custom: boolean,
     // The list of all dependencies (implicit or explicit).
     const allDirectDependencies = new Set<Resource>(explicitDirectDependencies);
 
-    const allDirectDependencyURNs = await getResourceURNs(explicitDirectDependencies);
+    const allDirectDependencyURNs = await getCustomResourceURNs(explicitDirectDependencies);
```
Comment (pgavlin) [inline sdk/nodejs/runtime/resource.ts]: Please change the name of this local to indicate that it only includes custom resources. Likewise below.

File: sdk/nodejs/resource.ts (hunk 4 above)
Comment (CyrusNajmabadi) [reply sdk/nodejs/resource.ts]: it helps ensure tat the member shows up as being used.  It's a little thing that bothers me when i'm editing the files.

File: sdk/nodejs/runtime/resource.ts (hunk 6 above)
Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/resource.ts]: done.

**Comment by CyrusNajmabadi** [comment]
@swgillespie @pgavlin PR updated.

I did change things slightly over from the initial PR.  Specifically, giving it more thought, i now include the referenced custom-resources (i.e. referenced through a ComponentResource) to the actual deps passed to the engine.  i.e. if you have:

```
//            comp1
//            /   \
//        cust1   comp2
//                /   \
//            cust2    cust3
//            /
//       cust4
```

And some resource 'dependsOn' 'comp1', then we will explicitly tell the engine that it actually `dependsOn: [cust1, cust2, cust3]`

This makes sense to me, and it does seem like the information that the engine would need to make good decisions.  If a resource really is dependent on a component, then that means it is actually dependent on those custom resources.

Also of note: teh dependencies we pass to the engine should now only be 'custom resources' (that was always in this PR).

**Comment by pgavlin** [comment]
> This makes sense to me, and it does seem like the information that the engine would need to make good decisions. If a resource really is dependent on a component, then that means it is actually dependent on those custom resources.

This worries me, as it makes it very easy for users to accidentally depend on resources that do not actually influence the properties of the dependent resource.

We have two sorts of dependencies: `dependOn` dependencies, which can be conservative, and per-property dependencies, which should not. It is okay for the former to be conservative because we only use that information to control the order in which resources are deleted. The latter, however, inform the set of resources that are deleted as a consequence of a delete-before-replace operation, and excess conservatism in this set can cause resources to be deleted even if it is not physically necessary to delete them.

I think it's okay for these extra custom resources to end up in the `dependsOn` set--the worst that will happen here is less available delete parallelism--but I do not think that these should be included in per-property dependencies.

**Comment by pgavlin** [comment]
It is also worth thinking about data vs. control dependencies. For example, it would be weird for a component resource to appear in the set of resources that an output depends on, but it may be very natural for a component to appear in a `dependsOn` list.

**Comment by CyrusNajmabadi** [comment]
Waiting on @pgavlin to make sure he's ok with the current approach.

**Comment by CyrusNajmabadi** [comment]
> I think it's okay for these extra custom resources to end up in the `dependsOn` set--the worst that will happen here is less available delete parallelism--but I do not think that these should be included in per-property dependencies

Wouldn't this only happen if you actually put a component resource in the 'property deps'?  Wouldn't this already be pretty rare/unlkely for someone to do?  And, if they did do this... would it necessarily be bad to say "that's a shorthand way of saying you depend on hte children of teh component?

**Comment by pgavlin** [comment]
> Wouldn't this only happen if you actually put a component resource in the 'property deps'? Wouldn't this already be pretty rare/unlkely for someone to do?

Depends on how you've written the code. Which is something I'm still looking at... :P

> And, if they did do this... would it necessarily be bad to say "that's a shorthand way of saying you depend on hte children of teh component?

It could be bad, yes. As per my earlier comment:

> [per-property dependencies], however, inform the set of resources that are deleted as a consequence of a delete-before-replace operation, and excess conservatism in this set can cause resources to be deleted even if it is not physically necessary to delete them.

**Comment by CyrusNajmabadi** [comment]
I guess i'm not getting it.  @swgillespie raised a good point in private discussions.  We'd like to think of Components as just being an alias/shorthand for just mentioning the explicit CustomResources.  i.e. you could generally think about them as being interchangable.  so, if you had written:

```new Whatever(name, { p: someComponent })``` that would effectively be equiv to something like:
```new Whatever(name, { p: [cust1, cust2, cust3] }```

if you wrote the latter, then you'd have the behavior you mentioned.  And, as such, it becomes less clear why you wouldn't get that behavior if you wrote the former.

GOing back to my bias toward consistency.  I feel like it would be more difficult to explain what Components are to users the more differin behavior we see from them in different locations.  If i say `dependsOn: someComponent` and that is equiv to `dependsOn: [cust1, cust2, cust3]`, then `p: someComponent` naturally feelsl ike it should behave the same way to me.

**Comment by CyrusNajmabadi** [comment]
> Depends on how you've written the code. Which is something I'm still looking at... :P

Could you expand on this?  First, components themselves never have prop-bags anymore.  So this is only about creating a custom resource, and passing it a props-bag that now references a Component.  

First, is there a realistic use case where that would happen?  Second, if it did happen, what do we want the semantics to even be of it.  If it's realistic, but the Component is *not* equivalent to its child-custom-resources, then what purpose does it even serve for the Component to be in the props-bag in the first place?  It wouldn't have any impact at all, and as such seems highly suspect.

--

This came up yesterday with discussions with @lukehoban .  We're willing to state that if something had practically zero meaning in the past, we're ok with it now taking on new and consistent meaning in the future.  This feels like one of those cases (unless i'm missing a core part of this that makes it meaningful).

**Comment by pgavlin** [comment]
I think that having a concrete example of the usage to talk through would be useful. @CyrusNajmabadi @lukehoban @swgillespie  do any you have one offhand? If not, I will put one together.

**Comment by CyrusNajmabadi** [comment]
Concretely, an example would be something like



```
awsx.Cluster (Component)
            \
             awsx.AutoScalingGroup (Component)
                                  \
                                  aws.cloudformation.Stack (Custom)
```

When making an awsx.Service, we want to be able to simply say that we 'dependsOn' that awsx.Cluster. And that will ensure that we cannot proceed until the actual `aws.cloudformation.Stack` is completed.

Right now, `awsx.Service` has to know very specifically the whitebox details of `awsx.Cluster` so that it can set up this dependsOn relationship.

**Comment by CyrusNajmabadi** [comment]
Closing in favor of https://github.com/pulumi/pulumi/pull/2509
